File: 2022/day15/solve.py
#!/usr/bin/env pypy3

from lib15 import *
from math import inf
import re
import sys

def SolvePart1(sensors, beacons, target_y):
  # Calculate horizontal segments intersecting target_y
  segments = []
  for x, y, r in sensors:
    dx = r - abs(y - target_y)
    if dx >= 0:
      segments.append((x - dx, x + dx))
  # Count total distance covered by segments, ignoring overlap
  segments.sort()
  covered = 0
  last_x = -inf
  for x1, x2 in segments:
    if last_x < x2:
      if last_x < x1:
        covered += x2 - x1 + 1
      else:
        covered += x2 - last_x
      last_x = x2
  # Remove beacons themselves from the count (they necessarily overlap)
  beacon_xs = set(x for x, y in beacons if y == target_y)
  return covered - len(beacon_xs)


def GeneratePart2Candidates(sensors, max_x, max_y):
  def IsCovered(bx, by):
    for sx, sy, r in sensors:
      if abs(sx - bx) + abs(sy - by) <= r:
        return True
    return False

  # The beacon must lie just outside the boundary of a sensor's range.
  # Unless the beacon lies on a boundary of the grid (0, 0)-(max_x,max_y)
  # (which seems unlikely), it must lie on the intersection of at least two
  # sensor's boundaries, so calculate these intersection points.
  diags1 = []  # x increasing, y increasing
  diags2 = []  # x increasing, y decreasing
  for x, y, r in sensors:
    diags1.append((x, y - r - 1, x + r + 1, y))
    diags1.append((x - r - 1, y, x, y + r + 1))
    diags2.append((x - r - 1, y, x, y - r - 1))
    diags2.append((x, y + r + 1, x + r + 1, y))
  candidates = set()
  for d1x1, d1y1, d1x2, d1y2 in diags1:
    for d2x1, d2y1, d2x2, d2y2 in diags2:
      # calculate intersection point via x-axis intersection
      x1 = (d1x1 - d1y1)
      x2 = (d2x1 + d2y1)
      d = x2 - x1
      if d % 2 != 0: continue  # not on integer point
      y = d // 2
      x = x1 + y
      assert x == x2 - y
      if (max(0, d1x1, d2x1) <= x <= min(max_x, d1x2, d2x2) and
          max(0, d1y1, d2y2) <= y <= min(max_y, d1y2, d2y1)):
        candidates.add((x, y))

  # Find the only candidate that is not covered by any sensors:
  for x, y in candidates:
    if not IsCovered(x, y):
      yield (x, y)

    # Check surrounding points too. This to guarantee uniqueness of the solution!
    for dx in range(-1, 2):
      for dy in range(-1, 2):
        if (dx != 0 or dy != 0) and not IsCovered(x + dx, y + dy):
          yield (x + dx, y + dy)

def SolvePart2(sensors, max_x, max_y):
  answer, = (4000000 * x + y for x, y in GeneratePart2Candidates(sensors, max_x, max_y))
  return answer

if __name__ == '__main__':
  coords = ReadCoords()
  sensors = MakeSensors(coords)
  beacons = MakeBeacons(coords)

  # Check that each sensor points to the nearest beacon
  for x, y, r in sensors:
    num_nearest = 0
    for bx, by in beacons:
      d = abs(x - bx) + abs(y - by)
      assert abs(x - bx) + abs(y - by) >= r
      num_nearest += d == r
    assert num_nearest == 1

  print(SolvePart1(sensors, beacons, 2000000))
  print(SolvePart2(sensors, 4000000, 4000000))

File: 2022/day15/solve.cc
// Advent of Code 2022 Day 14: Regolith Reservoir
// https://adventofcode.com/2022/day/14
//

#include <algorithm>
#include <cassert>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <iostream>
#include <string>
#include <vector>

// For memory map
// #include "sys/mman.h"
// #include "sys/stat.h"
// #include "unistd.h"

namespace {

constexpr int target_y = 2'000'000;
constexpr int multiply = 4'000'000;
constexpr int max_x    = 4'000'000;
constexpr int max_y    = 4'000'000;

struct Sensor {
  int x, y, r;
};

struct Line {
  int x1, y1, x2, y2;
};

bool IsCovered(const std::vector<Sensor> &sensors, int bx, int by) {
  // maybe binary search for viable range?
  for (auto [x, y, r] : sensors) {
    if (abs(x - bx) + abs(y - by) <= r) return true;
  }
  return false;
}

template <class T> void MakeUnique(std::vector<T> &v) {
  std::sort(v.begin(), v.end());
  v.erase(std::unique(v.begin(), v.end()), v.end());
}

int CountSegmentsSpan(std::vector<std::pair<int, int>> &segments) {
  std::sort(segments.begin(), segments.end());
  int last_x = std::numeric_limits<int>::min();
  int covered = 0;
  for (auto [x1, x2] : segments) {
    if (last_x < x2) {
      if (last_x < x1) {
        covered += x2 - x1 + 1;
      } else {
        covered += x2 - last_x;
      }
      last_x = x2;
    }
  }
  return covered;
}

uint64_t DelayMicros(
    std::chrono::steady_clock::time_point start,
    std::chrono::steady_clock::time_point finish) {
  return std::chrono::duration_cast<std::chrono::microseconds>(finish - start).count();
}

void CalculateDiagonals(
    const std::vector<Sensor> sensors,
    std::vector<Line> &diags1,
    std::vector<Line> &diags2) {
  for (auto [x, y, r] : sensors) {
    diags1.push_back(Line{x, y - r - 1, x + r + 1, y});
    diags1.push_back(Line{x - r - 1, y, x, y + r + 1});
    diags2.push_back(Line{x - r - 1, y, x, y - r - 1});
    diags2.push_back(Line{x, y + r + 1, x + r + 1, y});
  }
}

// This takes the most time. I could probably optimize it e.g. by partitioning
// the space into blocks and doing intersection within blocks.
void CalculateCandidates(
    const std::vector<Line> &diags1,
    const std::vector<Line> &diags2,
    std::vector<std::pair<int, int>> &candidates) {
  for (auto [d1x1, d1y1, d1x2, d1y2] : diags1) {
    int x_lo = std::max(d1x1, 0);
    int x_hi = std::min(d1x2, max_x);
    int y_lo = std::max(0, d1y1);
    int y_hi = std::min(d1y2, max_y);
    for (auto [d2x1, d2y1, d2x2, d2y2] : diags2) {
      // calculate intersection point via x-axis intersection
      int x1 = (d1x1 - d1y1);
      int x2 = (d2x1 + d2y1);
      int d = x2 - x1;
      if (d & 1) continue;  // not on integer point
      int y = d >> 1;
      int x = x1 + y;
      assert(x == x2 - y);
      if (x >= std::max(x_lo, d2x1) && x <= std::min(x_hi, d2x2) &&
          y >= std::max(y_lo, d2y2) && y <= std::min(y_hi, d2y1)) {
        candidates.push_back({x, y});
      }
    }
  }
  std::cerr << candidates.size() << " candidates\n";
  MakeUnique(candidates);
  std::cerr << candidates.size() << " unique candidates\n";
}

}  // namespace

int main() {
  auto time_start = std::chrono::steady_clock::now();

  // For part 1
  std::vector<int> beacon_xs;
  std::vector<std::pair<int, int>> segments;

  // For part 2
  std::vector<Sensor> sensors;

  // Read input.
  int x, y, bx, by;
  while (scanf("Sensor at x=%d, y=%d: closest beacon is at x=%d, y=%d\n", &x, &y, &bx, &by) == 4) {
    int r = abs(bx - x) + abs(by - y);

    // For part 1
    int dx = r - abs(y - target_y);
    if (dx >= 0) segments.push_back({x - dx, x + dx});
    if (by == target_y) beacon_xs.push_back(bx);

    // For part 2
    sensors.push_back(Sensor{x, y, r});
  }

  // Solve part 1.
  MakeUnique(beacon_xs);
  int answer1 = CountSegmentsSpan(segments) - beacon_xs.size();

  // Solve part 2.
  std::vector<Line> diags1;
  std::vector<Line> diags2;
  CalculateDiagonals(sensors, diags1, diags2);

  // I could speed this up by deduplicating while searching and stopping when I
  // find the first solution, but that would prevent me from using this solution
  // to check the correctness of my own data sets.
  std::vector<std::pair<int, int>> candidates;
  CalculateCandidates(diags1, diags2, candidates);

  std::chrono::steady_clock::time_point time_finish;
  long long answer2 = -1;
  for (auto [x, y] : candidates) {
    if (!IsCovered(sensors, x, y)) {
      // Solution found!
      if (answer2 != -1) {
        std::cerr << "Multiple solutions! "
            << (answer2 / multiply) << ',' << (answer2 % multiply) << " and "
             << x << ',' << y << std::endl;
        exit(1);
      }
      answer2 = (long long) x * multiply + y;
      time_finish = std::chrono::steady_clock::now();

      // Check for invalid data:
      for (int xx = x - 1; xx <= x + 1; ++xx) {
        for (int yy = y - 1; yy <= y + 1; ++yy) {
          if (xx != x || yy != y) {
            if (!IsCovered(sensors, xx, yy)) {
              std::cerr << "Multiple solutions! " << x << ',' << y << " and "
                  << xx << ',' << y << std::endl;
              exit(1);
            }
          }
        }
      }
    }
  }

  if (answer2 == -1) time_finish = std::chrono::steady_clock::now();

  std::cout << answer1 << std::endl;
  std::cout << answer2 << std::endl;

  std::cerr << "Time: " << DelayMicros(time_start, time_finish) << " us\n";
}
